### PR TITLE
Execute and & or conditions on multiple attributes

### PR DESCRIPTION
**Pull request by cendhu**

### CODE REVIEW

File: internal/queryexecutor/execute_conditions_test.go [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+
+		t.Run(tt.name, func(t *testing.T) {
+			keys, err := env.e.executeOR(dbName, tt.attrsConds)
+			require.NoError(t, err)
+			require.Equal(t, tt.expectedKeys, keys)
+		})
+	}
+}
+
+func TestExecute(t *testing.T) {
```
Comment (tock-ibm) [inline internal/queryexecutor/execute_conditions_test.go:344]: I am missing a test case for  `gt X` and `lt Y` where `X>Y`, expecting empty result

File: internal/queryexecutor/execute_conditions.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	startKey, err := json.Marshal(plan.startKey)
+	if err != nil {
+		return nil, err
+	}
+	endKey, err := json.Marshal(plan.endKey)
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (cendhu) [reply internal/queryexecutor/execute_conditions.go:133]: Done.

File: internal/queryexecutor/execute_conditions.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+	iter, err := e.db.GetIterator(stateindex.IndexDB(dbName), string(startKey), string(endKey))
```
[the remark this answers is not in the document]
Comment (cendhu) [reply internal/queryexecutor/execute_conditions.go:135]: Done.

File: internal/queryexecutor/execute_conditions.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	keys := make(map[string]bool)
+	for iter.Next() {
+		if iter.Error() != nil {
+			return nil, err
+		}
+
+		indexEntry := &stateindex.IndexEntry{}
+		if err := json.Unmarshal([]byte(iter.Key()), indexEntry); err != nil {
```
[the remark this answers is not in the document]
Comment (cendhu) [reply internal/queryexecutor/execute_conditions.go:150]: Done

File: internal/queryexecutor/execute_conditions.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	wg.Add(len(attrsConds))
+	errC := make(chan error, len(attrsConds))
+
+	for attr, conds := range attrsConds {
+		go func(attr string, conds *attributeTypeAndConditions) {
+			defer wg.Done()
+
+			keys, err := e.execute(dbName, attr, conds)
+			if err != nil {
+				errC <- err
```
[the remark this answers is not in the document]
Comment (cendhu) [reply internal/queryexecutor/execute_conditions.go:100]: Done.

File: internal/queryexecutor/execute_conditions_test.go (hunk 1 above)
Comment (cendhu) [reply internal/queryexecutor/execute_conditions_test.go:344]: Done

File: internal/queryexecutor/execute_conditions.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+func (e *WorldStateQueryExecutor) executeAllConditions(dbName string, attrsConds attributeToConditions) (map[string]map[string]bool, error) {
+	// Note that we simply create query plan for each condition and execute the same without
+	// performing any kind of query optimization. In the future, we can use statistics on number
+	// index entries per attribute and type to come up with complex query optimization methods.
+	attrKeys := &attributesKeys{
+		k: make(map[string]map[string]bool),
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(attrsConds))
```
[the remark this answers is not in the document]
Comment (cendhu) [reply internal/queryexecutor/execute_conditions.go:91]: Agree. Will be addressed by https://github.com/IBM-Blockchain/bcdb-server/issues/173

File: internal/queryexecutor/execute_conditions.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+	iter, err := e.db.GetIterator(stateindex.IndexDB(dbName), string(startKey), string(endKey))
+	if err != nil {
+		return nil, err
+	}
+	if iter.Error() != nil {
+		return nil, err
+	}
+
+	keys := make(map[string]bool)
+	for iter.Next() {
```
[the remark this answers is not in the document]
Comment (cendhu) [reply internal/queryexecutor/execute_conditions.go:144]: Good point. Will be addressed by https://github.com/IBM-Blockchain/bcdb-server/issues/174

File: internal/queryexecutor/execute_conditions.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+	m sync.Mutex
+}
+
+func (e *WorldStateQueryExecutor) executeAllConditions(dbName string, attrsConds attributeToConditions) (map[string]map[string]bool, error) {
+	// Note that we simply create query plan for each condition and execute the same without
+	// performing any kind of query optimization. In the future, we can use statistics on number
+	// index entries per attribute and type to come up with complex query optimization methods.
+	attrKeys := &attributesKeys{
+		k: make(map[string]map[string]bool),
+	}
```
[the remark this answers is not in the document]
Comment (cendhu) [reply internal/queryexecutor/execute_conditions.go:88]: Will be addressed by #174

**Review by gennadylaventman — APPROVED** [review]
